utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,187 @@
+/// 1. N must be equal to core::mem::sizeo_of::<Self>() or else it will
+///    cause panics.
+/// 2. The bit pattern [`NicheBytes::NICHE_BIT_PATTERN`] must not be incorrect as it would lead to
+///    weird behaviour.
+/// 3. The abstractions built on top of this trait must panic on an invalid N.
+/// 4. The abstractions built on this trait that use type punning must ensure that type being
+///    punned is [`ULE`].
+pub trait NicheBytes<const N: usize> {
+    const NICHE_BIT_PATTERN: [u8; N];
+}
+
+/// [`ULE`] type for [`NichedOption<U,N>`] where U implements [`NicheBytes`].
+/// The invalid bit pattern is used as the niche.
+///
+/// This uses 1 byte less than [`crate::ule::OptionULE<U>`] to represent [`NichedOption<U,N>`]
+/// which Derefs to [`Option<U>`].
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+///
+/// # Example
+///
+/// ```
+/// use zerovec::ZeroVec;
+/// use zerovec::ule::NichedOption;
+/// use core::num::NonZeroI8;
+///
+/// let bytes = &[0x00, 0x01, 0x02, 0x00 ];
+/// let zv_no: ZeroVec<NichedOption<NonZeroI8, 1>> =
+///         ZeroVec::parse_byte_slice(bytes).expect("Unable to parse as NichedOption.");
+///
+/// assert_eq!(zv_no.get(0).map(|e| e.0), Some(None));
+/// assert_eq!(zv_no.get(1).map(|e| e.0), Some(NonZeroI8::new(1)));
+/// assert_eq!(zv_no.get(2).map(|e| e.0), Some(NonZeroI8::new(2)));
+/// assert_eq!(zv_no.get(3).map(|e| e.0), Some(None));
+/// ```
+// Invariants:
+// The union stores [`NicheBytes::NICHE_BIT_PATTERN`] when None.
+// Any other bit pattern is a valid.
+#[repr(C)]
+pub union NichedOptionULE<U: NicheBytes<N> + ULE, const N: usize> {
+    /// Invariant: The value is `niche` only if the bytes equal NICHE_BIT_PATTERN.
+    niche: [u8; N],
+    /// Invariant: The value is `valid` if the `niche` field does not match NICHE_BIT_PATTERN.
+    valid: U,
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> NichedOptionULE<U, N> {
+    /// New NichedOptionULE<U, N> from Option<U>
+    pub fn new(opt: Option<U>) -> Self {
+        assert!(N == core::mem::size_of::<U>());
+        match opt {
+            Some(u) => Self { valid: u },
+            None => Self {
+                niche: <U as NicheBytes<N>>::NICHE_BIT_PATTERN,
+            },
+        }
+    }
+
+    /// Convert to an Option<U>
+    pub fn get(self) -> Option<U> {
+        // Safety: The union stores NICHE_BIT_PATTERN when None otherwise a valid U
+        unsafe {
+            if self.niche == <U as NicheBytes<N>>::NICHE_BIT_PATTERN {
+                None
+            } else {
+                Some(self.valid)
+            }
+        }
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> From<Option<U>> for NichedOptionULE<U, N> {
+    fn from(opt: Option<U>) -> Self {
+        Self::new(opt)
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> From<NichedOptionULE<U, N>> for Option<U> {
+    fn from(noule: NichedOptionULE<U, N>) -> Self {
+        NichedOptionULE::get(noule)
+    }
+}
```

Comment:
Do we need these? `AsULE` should be used to convert between ULE and non-ule types.

Reply:
Yeah these can be removed.

Reply:
We're not 100% perfect on getting the upgraded types from our ZeroVec APIs. Sometimes you end up with a ULE. Having an impl to convert at least `NichedOptionULE` to `Option` seems possibly useful.

It at least doesn't _hurt_ anything to have this.

Reply:
ok.